esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -6268,119 +6212,194 @@ void CWsDfuEx::exportRecordDefinitionBinaryType(const char *recordDefinition, Me
+        if (clusterType == NoCluster)
+             throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "DFUCreateFile: Cluster %s not found.", clusterName);
+
+        StringBuffer groupName;
+        getClusterGroupName(groupName, clusterName);
+        if (isEmptyString(groupName))
+             throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "DFUCreateFile: Group for cluster %s not found.", clusterName);
+        Owned<IGroup> group = queryNamedGroupStore().lookup(groupName.str());
+        if (!group)
+            throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "DFUCreateFile: Failed to get Group %s.", groupName.str());
+
         CDfsLogicalFileName lfn;
         lfn.set(fileName);
-        StringBuffer tempFileName = lfn.get();
+        StringBuffer normalizedFileName = lfn.get();
+
+        checkLogicalName(normalizedFileName, userDesc, false, true, false, nullptr);
+
+        StringBuffer tempFileName = normalizedFileName;
+        tempFileName.append(".").append(dfuCreateUniqId++); // avoid potential clash if >1 creating file. One will succeed at publish time.
         tempFileName.append(DFUFileCreate_FileNamePostfix);
 
-        StringBuffer jobId = requestBase.getJobId();
-        if (jobId.isEmpty())
-            jobId.appendf("Create %s on %s", tempFileName.str(), clusterName);
+        if (requestId.isEmpty())
+            requestId.appendf("Create %s on %s", normalizedFileName.str(), clusterName);
 
-        Owned<IFileDescriptor> fileDesc = createFileDescriptor(tempFileName, clusterTypeEx, groupName, group);
-        fileDesc->queryProperties().setProp("@job", jobId);
+        Owned<IFileDescriptor> fileDesc = createFileDescriptor(tempFileName, clusterTypeString(clusterType, false), groupName, group);
+        fileDesc->queryProperties().setProp("@job", requestId);
         if (!userId.isEmpty())
             fileDesc->queryProperties().setProp("@owner", userId);
+        fileDesc->queryProperties().setProp("ECL", recordDefinition);
 
-        Owned<IDistributedFile> file = queryDistributedFileDirectory().createNew(fileDesc);
-        file->setAccessed();
-        file->setECL(recordDefinition);
-        file->queryAttributes().setPropBin("_rtlType", layoutBin.length(), layoutBin.toByteArray());
-        file->queryAttributes().setPropInt64("@recordCount", 0);
-        file->queryAttributes().setPropInt64("@size", 0);
+        MemoryBuffer layoutBin;
+        exportRecordDefinitionBinaryType(recordDefinition, layoutBin);
+        if (0 == layoutBin.length())
+            throw makeStringExceptionV(ECLWATCH_INVALID_ECLRECDEF, "DFUCreateFile: Failed to parse ECL record definition: %s.", recordDefinition);
+        fileDesc->queryProperties().setPropBin("_rtlType", layoutBin.length(), layoutBin.toByteArray());
 
-        getFileAccessBeforePublish(context, tempFileName, clusterName, jobId, layoutBin, fileDesc, userDesc, requestBase, resp);
+        // NB: if file has copies on >1 cluster, they must share the same key
```

Comment:
Do we detect/report mismatched key in the situation described above? If not, will the user simply receive a failed file access?

Reply:
Yes, getFileDafilesrvConfiguration will generate an error if mismatches.